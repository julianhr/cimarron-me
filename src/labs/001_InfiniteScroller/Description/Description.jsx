import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import mdScrollerIntObs from './scroller_int_obs.md'
import mdScrollerContainerHeights from './scroller_container_height.md'
import mdScrolerSentinelClientRect from './scroller_sentinel_client_rect.md'
import GithubLogo from '~/components/library/GithubLogo'


export const SCROLLERS = {
  'intersectionObserver': mdScrollerIntObs,
  'containerScrollHeights': mdScrollerContainerHeights,
  'sentinelClientRect': mdScrolerSentinelClientRect,
}

const GITHUB_URL = "https://github.com/julianhr/cimarron.me/tree/master/src/labs/001_InfiniteScroller"

const Root = styled.section`
  padding: 0 30px 20px;
  width: 100%;

  ${props => props.theme.queries.from('md')} {
    padding: 40px 30px 20px;
  }
`

const P = styled.p`
  padding-top: 20px;
`

export class Description extends React.PureComponent {
  static propTypes = {
    scrollerType: PropTypes.string,
  }
  
  getDescription() {
    if (this.props.scrollerType in SCROLLERS) {
      return SCROLLERS[this.props.scrollerType]
    } else {
      return <React.Fragment />
    }
  }

  render() {
    return (
      <Root>
        <div
          dangerouslySetInnerHTML={{ __html: this.getDescription() }}
        />
        <P>
          source on <GithubLogo url={GITHUB_URL} styles={{ img: { marginTop: -4 } }} />
        </P>
      </Root>
    )
  }
}

const mapStateToProps = ({ scrollerType }) => (
  { scrollerType }
)

export default connect(mapStateToProps)(Description)
