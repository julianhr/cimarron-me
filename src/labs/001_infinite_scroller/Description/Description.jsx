import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import mdScrollerIntObs from './scroller_int_obs.md'
import mdScrollerContainerHeights from './scroller_container_height.md'
import mdScrolerSentinelClientRect from './scroller_sentinel_client_rect.md'
import GithubLogo from '~/components/library/GithubLogo'


const Root = styled.section`
  grid-area: b1;
  padding: 0 30px 20px;
  width: 100%;

  ${props => props.theme.queries.from('md')} {
    grid-row: 2 / span 1;
    grid-column: 1 / span 1
  }
`

const P = styled.p`
  padding-top: 20px;
`

class Description extends React.PureComponent {
  static propTypes = {
    scrollerType: PropTypes.string,
  }
  
  getDescription() {
    switch (this.props.scrollerType) {
      case 'intersectionObserver':
        return mdScrollerIntObs
      case 'containerScrollHeights':
        return mdScrollerContainerHeights
      case 'sentinelClientRect':
        return mdScrolerSentinelClientRect
    }
  }

  render() {
    const githubUrl = "https://github.com/julianhr/cimarron-me/tree/master/src/labs/001_infinite_scroller"

    return (
      <Root>
        <div
          dangerouslySetInnerHTML={{ __html: this.getDescription() }}
        />
        <P>
          source on <GithubLogo url={githubUrl} styles={{ img: { marginTop: -4 } }} />
        </P>
      </Root>
    )
  }
}

const mapStateToProps = ({ scrollerType }) => (
  { scrollerType }
)

export default connect(mapStateToProps)(Description)
