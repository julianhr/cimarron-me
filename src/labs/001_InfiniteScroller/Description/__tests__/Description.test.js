import React from 'react'
import { create } from 'react-test-renderer'

import { Description, SCROLLERS } from '../Description'
import MockInfiniteScrollerApp from '../../__tests__/__mocks__/MockInfiniteScrollerApp'


describe('Description', () => {
  describe('snapshots', () => {
    it('with valid md file', () => {
      const mdString = Object.keys(SCROLLERS)[0]
  
      const wrapper = create(
        <MockInfiniteScrollerApp>
          <Description scrollerType={mdString} />
        </MockInfiniteScrollerApp>
      )
      expect(wrapper).toMatchSnapshot('valid md file')
    })
  })

  it('with invalid md file', () => {
      const wrapper = create(
        <MockInfiniteScrollerApp>
          <Description scrollerType={'invalid'} />
        </MockInfiniteScrollerApp>
      )
      expect(wrapper).toMatchSnapshot('invalid md file')
      
      wrapper.update(
        <MockInfiniteScrollerApp>
          <Description />
        </MockInfiniteScrollerApp>
      )
      expect(wrapper).toMatchSnapshot('no md file')
  })

  it('with invalid md file', () => {
    for (let [key, _] of Object.entries(SCROLLERS)) {
      const wrapper = create(
        <MockInfiniteScrollerApp>
          <Description scrollerType={key} />
        </MockInfiniteScrollerApp>
      )

      const nodes = wrapper.root.findAll(node => node.props.hasOwnProperty('dangerouslySetInnerHTML'))
      expect(nodes).toHaveLength(1)
    }
  })
})
