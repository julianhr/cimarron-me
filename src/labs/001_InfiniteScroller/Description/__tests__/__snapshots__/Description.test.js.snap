// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Description snapshots with valid md file: valid md file 1`] = `
<section
  className="css-9l38gp-Root eyj91y30"
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "test-file-stub",
      }
    }
  />
  <p
    className="css-1vm6kb8-P eyj91y31"
  >
    source on 
    <span>
      <a
        href="https://github.com/julianhr/cimarron.me/tree/master/src/labs/001_InfiniteScroller"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="GitHub"
          className="css-1inomk8-Img-GithubLogo epzvwn20"
          src="test-file-stub"
        />
      </a>
    </span>
  </p>
</section>
`;

exports[`Description with invalid md file: invalid md file 1`] = `
<section
  className="css-9l38gp-Root eyj91y30"
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": <React.Fragment />,
      }
    }
  />
  <p
    className="css-1vm6kb8-P eyj91y31"
  >
    source on 
    <span>
      <a
        href="https://github.com/julianhr/cimarron.me/tree/master/src/labs/001_InfiniteScroller"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="GitHub"
          className="css-1inomk8-Img-GithubLogo epzvwn20"
          src="test-file-stub"
        />
      </a>
    </span>
  </p>
</section>
`;

exports[`Description with invalid md file: no md file 1`] = `
<section
  className="css-9l38gp-Root eyj91y30"
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": <React.Fragment />,
      }
    }
  />
  <p
    className="css-1vm6kb8-P eyj91y31"
  >
    source on 
    <span>
      <a
        href="https://github.com/julianhr/cimarron.me/tree/master/src/labs/001_InfiniteScroller"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="GitHub"
          className="css-1inomk8-Img-GithubLogo epzvwn20"
          src="test-file-stub"
        />
      </a>
    </span>
  </p>
</section>
`;
